jax/experimental/sparse/bcoo.py

```diff
@@ -1375,6 +1351,101 @@ def _bcoo_spdot_general_jvp(primals, tangents, **kwds):
+# Utility to sort the indices of a BCOO representation. This primitive
+# does not support deduplication or removing of zeros; see bcoo_sum_duplicates.
+
+bcoo_sort_indices_p = core.Primitive("bcoo_sort_indices")
+bcoo_sort_indices_p.multiple_results = True
+
+def bcoo_sort_indices(mat):
+  """Sort indices of a BCOO array, and optionally sum duplicates & eliminate zeros.
+
+  Args:
+    mat : BCOO array
+
+  Returns:
+    mat_out : BCOO array with sorted indices.
+  """
+  data, indices = bcoo_sort_indices_p.bind(*mat._bufs, spinfo=mat._info)
+  return BCOO((data, indices), shape=mat.shape)
+
+@bcoo_sort_indices_p.def_impl
+def _bcoo_sort_indices_impl(data, indices, *, spinfo):
+  props = _validate_bcoo(data, indices, spinfo.shape)
+  if props.n_sparse == 0:
+    return data, indices
+  f = _bcoo_sort_indices_unbatched
+  for _ in range(props.n_batch):
+    f = vmap(f)
+  indices, perm = f(indices)
+  permute = lambda d, p: d[p]
+  for _ in range(props.n_batch):
+    permute = broadcasting_vmap(permute)
+  data = permute(data, perm)
+  return data, indices
+
+def _bcoo_sort_indices_unbatched(indices):
+  # sort indices without summing duplicates
+  nse, N = indices.shape
+  idx_cols = (indices[:, i] for i in range(N))
+  *indices, perm = lax.sort((*idx_cols, lax.iota(indices.dtype, nse)), num_keys=N)
+  return jnp.column_stack(indices), perm
+
+@bcoo_sort_indices_p.def_abstract_eval
+def _bcoo_sort_indices_abstract_eval(data, indices, *, spinfo):
+  props = _validate_bcoo(data, indices, spinfo.shape)
+  if props.n_sparse == 0:
+    return data, indices
+  data_out = core.ShapedArray(
+    (*map(max, indices.shape[:props.n_batch], data.shape[:props.n_batch]),
+     props.nse, *data.shape[props.n_batch + 1:]), data.dtype, weak_type=data.weak_type)
+  return data_out, indices
+
+def _bcoo_sort_indices_batching_rule(batched_args, batch_dims, *, spinfo):
+  data, indices = batched_args
+  if any(b not in [0, None] for b in batch_dims):
+    raise NotImplementedError(f"batch_dims={batch_dims}. Only 0 and None are supported.")
+  if batch_dims[0] is None:
+    data = data[None, ...]
+  if batch_dims[1] is None:
+    indices = indices[None, ...]
+  new_spinfo = BCOOInfo(shape=(max(data.shape[0], indices.shape[0]), *spinfo.shape))
+  return bcoo_sort_indices_p.bind(data, indices, spinfo=new_spinfo), (0, 0)
+
+def _bcoo_sort_indices_jvp(primals, tangents, *, spinfo):
+  props = _validate_bcoo(*primals, spinfo.shape)
+  if props.n_sparse == 0:
+    return primals, tangents
+
+  data, indices = primals
+  data_dot, _ = tangents
+  f = _bcoo_sort_indices_unbatched
+  for _ in range(props.n_batch):
+    f = broadcasting_vmap(f)
+  indices_out, perm = f(indices)
+  permute = lambda d, p: d[p]
+  for _ in range(props.n_batch):
+    permute = broadcasting_vmap(permute)
+  data_out = permute(data, perm)
+
+  indices_dot_out = ad.Zero.from_value(indices)
+  data_dot_out = ad.Zero.from_value(data_out) if type(data_dot) is ad.Zero else permute(data_dot, perm)
```

Comment:
Looks like this idea worked out!

Reply:
Yeah, returning the permutation was the key